File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        Queue<int[]> que = new LinkedList<>();
        ArrayList<Integer>[] graph = new ArrayList[n + 1];
        int[] visited = new int[n + 1];

        que.offer(new int[]{1, 1});

        for (int[] e : edge) {
            if (graph[e[0]] == null) {
                graph[e[0]] = new ArrayList<>();
            }

            if (graph[e[1]] == null) {
                graph[e[1]] = new ArrayList<>();
            }

            graph[e[0]].add(e[1]);
            graph[e[1]].add(e[0]);
        }

        visited[1] = 1;

        while (!que.isEmpty()) {
            int[] now = que.poll();

            for (int g : graph[now[0]]) {
                if (visited[g] == 0) {
                    que.offer(new int[]{g, now[1] + 1});
                    visited[g] = now[1] + 1;
                }
            }
        }

        int max = Arrays.stream(visited).max().getAsInt();
        int cnt = 0;

        for (int v : visited) {
            if (v == max) {
                cnt++;
            }
        }

        return cnt;
    }
}

File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.kt
import java.util.*

class Solution {
    fun solution(n: Int, edge: Array<IntArray>): Int {
        val visited = IntArray(n + 1) { -1 }
        val que: Queue<Pair<Int, Int>> = LinkedList()
        val graph = MutableList(n + 1) { arrayListOf<Int>() }

        visited[1] = 0
        que.add(Pair(1, 0))
        edge.forEach {
            graph[it[0]].add(it[1])
            graph[it[1]].add(it[0])
        }

        while (que.isNotEmpty()) {
            val (node, cnt) = que.poll()

            graph[node].forEach {
                if (visited[it] == -1) {
                    visited[it] = cnt + 1
                    que.add(Pair(it, cnt + 1))
                }
            }
        }
    
        return visited.count { it == visited.maxOrNull() }
    }
}